### PR TITLE
add follower following

### PR DESCRIPTION
**Pull request by gravidesaldrin**
HI Sir. This is the PR for Following and Follower function. Please Review. Thanks.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -37,6 +41,16 @@ def update
     end
   end
 
+  def following
```
Comment (MuhammadTamzid) [inline app/controllers/users_controller.rb]: RESTFull

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -37,6 +41,16 @@ def update
...
+    @users = @user.following.paginate page: params[:page]
+    render :show_follow
+  end
+
+  def followers
```
Comment (MuhammadTamzid) [inline app/controllers/users_controller.rb]: RESTFull

File: app/models/relationship.rb [hunk 3]
```diff
@@ -1,4 +1,7 @@
 class Relationship < ActiveRecord::Base
   belongs_to :follower, class_name: User.name
   belongs_to :followed, class_name: User.name
+
+  validates :follower_id, presence: true
```
Comment (MuhammadTamzid) [inline app/models/relationship.rb]: validates :follower, presence: true

File: app/models/relationship.rb [hunk 4]
```diff
@@ -1,4 +1,7 @@
...
+  validates :followed_id, presence: true
```
Comment (MuhammadTamzid) [inline app/models/relationship.rb]: validates :followed, presence: true

**Comment by gravidesaldrin** [comment]
Ok Sir. The following and followers page will be in the next PR, because it is already 15 files. The link for following and follower in the app/views/shared_pages/_sidebar.html.erb is still "#". Please review. Thank You

**Comment by gravidesaldrin** [comment]
Ok Sir. The following and followers page will be in the next PR, because it is already 15 files. The link for following and follower in the app/views/shared_pages/_sidebar.html.erb is still "#". Please review. Thank You

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,10 +1,14 @@
...
   before_action :correct_user, only: [:edit, :update]
-  before_action :find_user, except: [:index, :new, :create]
+  before_action :find_user, except: [:index, :new, :create, :following, :followers]
 
   def index
-    @users = User.paginate(page: params[:page])
+    if current_user.admin?
+      @users = User.paginate page: params[:page]
+    else
+      @users = User.where(admin: false).paginate page: params[:page]
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: move where to scope in model

File: config/routes.rb [hunk 6]
```diff
@@ -9,6 +9,12 @@
   delete "logout" => "sessions#destroy"
 
   resources :users, only: [:index, :show, :new, :create, :edit, :update]
+  resources :users do
+    member do
+      get :following, :followers
```
Comment (dieunb) [inline config/routes.rb]: where are they?

**Comment by gravidesaldrin** [comment]
Ok Sir. I remove the :following and :followers path in the config/routes.rb, then put not admins in index of users. Please review my PR. Thank You

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@ class UsersController < ApplicationController
   before_action :find_user, except: [:index, :new, :create]
 
   def index
-    @users = User.paginate(page: params[:page])
+    @users = User.where(admin: false).paginate page: params[:page]
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: move `where` into `scope` in model, you can search `rails scope`

**Comment by gravidesaldrin** [comment]
I add can_follow scope in user.rb . Please review Sir. Thanks

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -4,7 +4,7 @@ class UsersController < ApplicationController
   before_action :find_user, except: [:index, :new, :create]
 
   def index
-    @users = User.paginate(page: params[:page])
+    @users = User.can_follow.paginate page: params[:page]
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: where is `can_follow`?

File: app/models/user.rb [hunk 9]
```diff
@@ -23,6 +23,8 @@ class User < ActiveRecord::Base
   before_save :downcase_email
   before_create :create_activation_digest
 
+  scope :can_follow, -> {where admin: false}
```
Comment (dieunb) [inline app/models/user.rb]: bad naming. This naming isn't show that mean!

**Pull request merged**